source/extensions/listener_managers/listener_manager/fcds_impl.cc

```diff
@@ -0,0 +1,227 @@
+                              const std::string&) {
+
+  std::vector<const envoy::config::listener::v3::FilterChain*> removed_filter_chains;
+  std::vector<const envoy::config::listener::v3::FilterChain*> new_filter_chains;
+  std::string message;
+
+  TRY_ASSERT_MAIN_THREAD 
+    processDeletedFilterChains(removed_resources, removed_filter_chains);
+    processAddedOrUpdatedFilterChains(added_resources, removed_filter_chains, new_filter_chains);
+
+    // Finally invoke the deletion call
+    filter_chain_manager_.removeFilterChains(removed_filter_chains);
+
+    // Finally invoke the addition call
+    absl::Span<const envoy::config::listener::v3::FilterChain* const> filter_chain_span = new_filter_chains;
+    filter_chain_manager_.addFilterChains(filter_chain_span, nullptr, *fc_builder_,
+                                        filter_chain_manager_, true);
+
+    // Drain the filter chains which were deleted
+    filter_chain_manager_.startDrainingSequenceForListenerFilterChains();
+  END_TRY
+  catch (const EnvoyException& e)
+  {
+    absl::StrAppend(&message, e.what(), "\n");
+
+  }
+
+  local_init_target_.ready();
+
+  if (!message.empty()) {
+    throw EnvoyException(fmt::format("Error adding/updating filter chain(s) : {}\n", message));
+  }
+}
+
+void FcdsApi::onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reason,
+                                    const EnvoyException*) {
+  ASSERT(Envoy::Config::ConfigUpdateFailureReason::ConnectionFailure != reason);
+  // We need to allow server startup to continue, even if we have a bad
+  // config.
+  local_init_target_.ready();
+}
+
+bool FcdsApi::validateUpdateSize(int num_resources) {
+  if (num_resources == 0) {
+    ENVOY_LOG(debug, "Missing Filter Chain for fcds config block = {}", fcds_cfg_name_);
+    local_init_target_.ready();
+    return false;
+  }
+
+  ENVOY_LOG(debug, "fcds: received {} filter chain for fcds config block = {}", num_resources, fcds_cfg_name_);
+  return true;
+}
+
+
+// The deleted/draining FC list are constructed here.
+void FcdsApi::processDeletedFilterChains(const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+  std::vector<const envoy::config::listener::v3::FilterChain*>& removed_filter_chains) {
+  for (const std::string& filter_chain_name : removed_resources) {
+    markFilterChainForDeletion(filter_chain_name, removed_filter_chains);
+    ENVOY_LOG(debug, "fcds: removing filter chain from FCM: fcds config block name= {} filter chain={}", 
+        fcds_cfg_name_, filter_chain_name);
+  }
+}
+
+// The updated/added FC list are constructed here.
+void FcdsApi::processAddedOrUpdatedFilterChains(const std::vector<Envoy::Config::DecodedResourceRef>& added_resources,
+  std::vector<const envoy::config::listener::v3::FilterChain*>& updated_filter_chains, 
+  std::vector<const envoy::config::listener::v3::FilterChain*>& new_filter_chains) {
+  
+  std::vector<std::string> seen_filter_chains; 
```

Comment:
Probably should be flat_hash_set and not vector.

Reply:
Fixed